go/metaparticle/metaparticle.go

```diff
@@ -0,0 +1,168 @@
+package metaparticle
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"strings"
+)
+
+// Runtime represents a place where to run the containers (Docker, rkt, etc)
+type Runtime struct {
+	// must return an integer >= 1
+	Replicas int32
+
+	// must return an integer >= 0
+	Shards int32
+
+	// TODO: find out what this does
+	URLShardPattern string
+
+	// The name of the executor (e.g.: docker)
+	Executor string
+
+	// Returns the ports that the service exposes
+	Ports []int32
+
+	// Returns whether the service wants a public IP when deployed (usually involves the creation of a load balancer)
+	PublicAddress bool
+}
+
+type Package struct {
+	Repository string
+	Verbose    bool
+	Quiet      bool
+	Builder    string
+	Publish    bool
+}
+
+// Executor implementors are container platforms where the containers can be deployed (e.g. Azure, GCP)
+type Executor interface {
+	Run(image string, name string, config *Runtime, stdout io.Writer, stderr io.Writer) error
+	Logs(name string, stdout io.Writer, stderr io.Writer) error
+	Cancel(name string) error
+}
+
+// Builder is the interface that wraps the container runtime's build and push methods
+type Builder interface {
+	Build(dir string, image string, stdout io.Writer, stderr io.Writer) error
+	Push(image string, stdout io.Writer, stderr io.Writer) error
+}
+
+func inDockerContainer() bool {
+	if os.Getenv("METAPARTICLE_IN_CONTAINER") == "true" {
+		return true
+	}
+	b, err := ioutil.ReadFile("/proc/1/cgroup")
+
+	if err != nil {
+		return false
+	}
+	s := string(b)
+	return strings.Contains(s, "docker") || strings.Contains(s, "kubepods")
+}
+
+func executorFromRuntime(r *Runtime) (Executor, error) {
+	if r == nil {
+		return &DockerImpl{}, nil
+	}
+	switch r.Executor {
+	case "docker":
+		return &DockerImpl{}, nil
+	case "aci":
+		// TODO: find a way to parameterize the name of the resource group
+		return &ACIExecutor{"test"}, nil
+	case "metaparticle":
+		return &MetaparticleExecutor{}, nil
+	default:
+		return nil, fmt.Errorf("Unknown executor: %s", r.Executor)
+	}
+}
+
+func builderFromPackage(pkg *Package) (Builder, error) {
+	if pkg == nil {
+		return &DockerImpl{}, nil
+	}
+	switch pkg.Builder {
+	case "docker":
+		return &DockerImpl{}, nil
+	default:
+		return nil, fmt.Errorf("Unknown builder: %s", pkg.Builder)
+	}
+}
+
+func writeDockerfile(name string) error {
+	contents := `FROM golang:1.9
```

Comment:
Rather than doing it this way, I thought that the right way to do it would be to pull the static binary from `argv[0]` and just add that to a simple image.

Of course that will only work on Linux, so that's a bit of a limitation, but maybe we can take a `go build ...` directive as part of the configuration.

I'm not sure that we can assume a specific go project layout.